### PR TITLE
[FLINK-7608][metric] Refactor latency statistics metric

### PR DESCRIPTION
**Pull request by zentol**
This PR is an extension of #4826.

The additions addresses issues that I found while trying it out. Some of these unfortunately may be rather controversial, hence why I'm curious what others think.

#### 1)
The first (and probably less-controversial) addition is to register the metric on the job level with user-defined variables for proper integration with tag-based reporters. As described in #4826 the latency metric doesn't fit as a operator/task metric since it describes a relation between 2 operators.

#### 2)
The second problem i found was that the latency marker uses `StreamConfig#getVertexID` to retrieve the source/operator vertexID. This is however not the actual runtime ID, but just a counter that is incremented when the job is generated, and thus utterly pointless and impossible to correlate with a job's operators. 
This change requires serializing 2 longs instead of an int along with creating an `OperatorID` during deserialization, making the latency measuring more costly.

#### 3)
The third and last change that I made is removing the case distinction between operators and sinks. Previously, operators would ignore the source subtask index and tread all subtasks equally. Only sinks measured latency separately for the source subtask.
Despite the performance benefits I find this design rather questionable:
* latency metric names for operators&sinks are inconsistent, and there is in fact no way to tell whether something is an operator or sink without access to the job source code which means to query the metric you have to resort to trial-and-error
* there are operators with sink semantics (like the write-ahead sink), so we aren't consistent in terms of "all sinks have detailed latency metrics" and are leaking implementation details to the outside
* considering sink latency metrics as more important is rather arbitrary and may very well not fit a user's requirements

Instead, we may want to think about is adding a separate switch for detailed latency metrics.

/cc @aljoscha @rmetzger @yew1eb

### CODE REVIEW

**Comment by yew1eb** [comment]
@zentol :+1:
**1)**,  I have no opinion.
**2), 3)**, Replace `operatorID` with `operatorName`, will be better?

@aljoscha @rmetzger   what are your opinions?

**Comment by zentol** [comment]
@yew1eb Are you suggesting to store the operatorName in the latency marker?

**Comment by yew1eb** [comment]
Yes, in our production environment, we report and store all metrics to an external Time series database for alarm and visual presentation.

When the job is started, we will store the edge structure of the job's logical plan: eg. <Unique_job_source_id, jobName, operatorA, operatorB>.

For the latency metrics, we hope to include the operatorName instead of the operatorID in the tags. because operatorID will change if the job redeploys, the history's latency metrics data for the job will be hard to match and no latency statistics will be observed in our visualization panel.

**Comment by aljoscha** [comment]
@yew1eb by `operatorName`, do you mean the `name()` or `uid()`, I think both of these could make sense.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java [hunk 1]
```diff
@@ -194,14 +190,20 @@ public void setup(StreamTask<?, ?> containingTask, StreamConfig config, Output<S
...
-			historySize = MetricOptions.LATENCY_HISTORY_SIZE.defaultValue();
+
+		try {
+			Configuration taskManagerConfig = container.getEnvironment().getTaskManagerInfo().getConfiguration();
+			int historySize = taskManagerConfig.getInteger(MetricOptions.LATENCY_HISTORY_SIZE);
+			if (historySize <= 0) {
+				LOG.warn("{} has been set to a value equal or below 0: {}. Using default.", MetricOptions.LATENCY_HISTORY_SIZE, historySize);
+				historySize = MetricOptions.LATENCY_HISTORY_SIZE.defaultValue();
+			}
+			this.latencyStats = new LatencyStats(this.metrics.parent().parent().addGroup("latency"), historySize, container.getIndexInSubtaskGroup(), getOperatorID());
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java]: I guess what we could do add another method to the `TaskManagerJobMetricGroup` for creating the latency metric group? With that, modifications to the metric group structure will result in a compile error if `parent().parent()` does not return us to the job metric group.

**Comment by zentol** [comment]
I'm wary about using the operator name (or _any_ variant that inexplicably exists) as that significantly increases the size of watermarks.

I see the use-case, and ideally I would like to have both the operator name and id to both have a unique metric by default while supporting static names across jobs, but i don't see a way to do that without either a) blowing up watermarks even more or b) including an index of all operators (id+name) in the TDD and making that accessible.

**Comment by aljoscha** [comment]
I'd say we merge without operator names/operator id and leave that as a follow-up, issue. To at least make it possible to get latency metrics. WDYT?

Btw, with this, can we now show the latency metrics in the Flink web dashboard?

**Comment by zentol** [comment]
yes the display will now work.

**Comment by zentol** [comment]
I've rebased the branch and did one more change:
```
this.latencyStats = new LatencyStats(this.metrics.parent().parent().addGroup("latency"), historySize, container.getIndexInSubtaskGroup(), getOperatorID());
```
===>
```
TaskManagerJobMetricGroup jobMetricGroup = this.metrics.parent().parent();
this.latencyStats = new LatencyStats(jobMetricGroup.addGroup("latency"), historySize, container.getIndexInSubtaskGroup(), getOperatorID());
```

**Comment by aljoscha** [comment]
So this ensures that the parent is actually a `TaskManagerJobMetricGroup`, right?

This LGTM for me, now. Could you open a follow-up issue for the operator-name thing?

**Comment by zentol** [comment]
Correction: We still can't display them in the UI since we have no tab for job metrics.

**Comment by zentol** [comment]
merging.

**Pull request closed without merging**